[LLVM][IR] Add support for vector ConstantInt/FP to ConstandFolding:FoldBitCast.

paulwalker-arm opened the pull request.

paulwalker-arm committed:
[LLVM][IR] Add support for vector ConstantInt/FP to ConstandFolding:FoldBitCast.

david-arm commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -151,8 +151,14 @@ Constant *FoldBitCast(Constant *C, Type *DestTy, const DataLayout &DL) {
   // If this is a bitcast from constant vector -> vector, fold it.
-  if (!isa<ConstantDataVector>(C) && !isa<ConstantVector>(C))
+  if (!isa<ConstantDataVector>(C) && !isa<ConstantVector>(C) &&
+      !isa<ConstantInt>(C) && !isa<ConstantFP>(C))
```
Refers to line 206: Do we need to add ConstantInt and ConstantFP checks here too?

paulwalker-arm replied:
I was going to say no because it should be safe to not perform anymore transformations.  However, if I'm reading it correctly I think that block of code is just broken upstream?

For the case where constant folding does not happen it is effectively returning `ConstantExpr::getBitCast(C, SrcAsIntTy)` and by this point we already know `SrcAsIntTy != DestVTy`?

paulwalker-arm committed:
Prevent bogus return when constant fold does not happen.

Rather then return the correct result I believe we do not hit this
issue because the input has been restricted to constants that we
know can be constant folded and so I've replace the broken code with
an assert, which is in keeping with the other code in this area that
requires constant folding to work.

paulwalker-arm committed:
remove ConstantFP from assert given we know SrcIVTy is an integer

david-arm reviewed (approved):
LGTM!
